### ISSUE TITLE
Android Native client results inconsistent across multiple runs

### ISSUE THREAD

**Issue by rpratesh** [mozilla/DeepSpeech#2125]
- **Have I written custom code (as opposed to running examples on an unmodified clone of the repository)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Android P
- **TensorFlow installed from (our builds, or upstream TensorFlow)**: Your builds
- **TensorFlow version (use command below)**: Not applicable 
- **Python version**: Not applicable
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: Not Applicable
- **GPU model and memory**: Not Applicable
- **Exact command to reproduce**: 

Trying to run DeepSpeech 0.4.1 native_client on Android P platform with tflite model of DeepSpeech 0.4.1 ckpts.

When the client is run multiple times on same set of wav files the results are inconsistent.

In detail,
we have a folder with set of, say, 4 wav files, each copy of same one wav file.

Now if we run
`./deepspeech --model /docker_files/checkpoints_0_4_1_mozilla/output_graph_original.tflite --alphabet /docker_files/lm_own_v3/alphabet.txt --audio /docker_files/wavs/wav1/`

we get 

```
> > /data/local/wav1//call_9890278956_MAIN.wav
>  gol nin ight nines yelooked to seven eight nine five six
> > /data/local/wav1//call_9890278956_COPY1.wav
>  gol nine eight nine elook two seven eight nine five six
> > /data/local/wav1//call_9890278956_COPY2.wav
>  gol nine eight nine elo two seven eight nine five six
> > /data/local/wav1//call_9890278956_COPY3.wav
>  gol nine eight nine zelo two seven eight nine five six
```
Note that we are giving the path of folder that contains all the four wav files. The results are inconsistent even if we run with LM and trie and with different set of wav files.

While if we take each wav file individually and run the same above command, the output is consistent. It's always
`gol nine ight names yelooked to seven eight nine fivt six`

NOTE: We didn't find this inconsistency issue when working on Linux native client using .pb files and the same above wav files

**Comment by lissyx** [comment]
@rpratesh It seems you have a different alphabet. Can you reproduce with exactly our model ? With `LDC93S1` sample ?

TFLite model has been measured with a little more WER compared to pure protocolbuffer (approximately +2, from 8.5% WER to 10.2%), and the C++ `deepspeech` sample is not really intended for final use, it's possible the code is buggy when running on multiple WAVs.

**Comment by lissyx** [comment]
> **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Android P

Can you share more details on the platform ? I don't recall hitting any issue on Android 8.0 and Android 9.0, respectively on Xperia XZ and Pixel 2 devices.

**Comment by lissyx** [comment]
We also did quite some changes to the codebase regarding TFLite since v0.4.1, that may impact behavior as well, it would be great if you could try and verify this with current master on a smaller model (LDC93S1 for example).

**Comment by rpratesh** [comment]
> > **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Android P
> 
> Can you share more details on the platform ? I don't recall hitting any issue on Android 8.0 and Android 9.0, respectively on Xperia XZ and Pixel 2 devices.

We are using Qualcomm Snapdragon 820 development board. Also, may I know if u tried DS041 ckpts based .tflite model on the above platforms over multiple iterations and still got the consistent results.

> @rpratesh It seems you have a different alphabet. Can you reproduce with exactly our model ? With `LDC93S1` sample ?

The alphabet.txt is the same English alphabet.txt as shared by Deepspech team.
We are getting consistent result if we tried with .tflite model of Deepspeech with 494 n_hidden trained on LDC93S1, but with DS0.4.1 ckpts based .tflite model, it's still inconsistent with 0.4.1 native client.

Hopefully 

> The C++ `deepspeech` sample is not really intended for final use, it's possible the code is buggy when running on multiple WAVs.

If the C++ deepspeech code is buggy, then wouldn't be the native_client results on Linux also give inconsistent results (while I am getting consistent results using Linux native client)

**Comment by lissyx** [comment]
> We are using Qualcomm Snapdragon 820 development board. Also, may I know if u tried DS041 ckpts based .tflite model on the above platforms over multiple iterations and still got the consistent results.

We only really care about `libdeepspeech` through JNI so no.

> If the C++ deepspeech code is buggy, then wouldn't be the native_client results on Linux also give inconsistent results (while I am getting consistent results using Linux native client)

Same answer, the C++ code is here just for some testing facilites, it's also advertised not to rely on it.

**Comment by lissyx** [comment]
> We are getting consistent result if we tried with .tflite model of Deepspeech with 494 n_hidden trained on LDC93S1, but with DS0.4.1 ckpts based .tflite model, it's still inconsistent with 0.4.1 native client.

Ok, I would really need that you re-test from Java level. I do insist, TFLite on 0.4.1 is really new, don't be surprised too much if it's not perfect. Please test with current master as well.

**Comment by rpratesh** [comment]
> We only really care about `libdeepspeech` through JNI so no.

So you suggest me to use libdeepspeech through JNI instead of C++ deepspeech client ? And will v0.4.1 be good for that?

> Ok, I would really need that you re-test from Java level. I do insist, TFLite on 0.4.1 is really new, don't be surprised too much if it's not perfect. Please test with current master as well.

I can test with current master but then that's lots of change , mainly in terms of tensorflow version. Trying to get this resolved with 0.4.1 release itself.

**Comment by lissyx** [comment]
> So you suggest me to use libdeepspeech through JNI instead of C++ deepspeech client ? And will v0.4.1 be good for that?

Yes, that's what i'm suggesting. If the issue is within TFLite itself, it won't help, but at least you should be at some point in a stack where foundations are better.

> I can test with current master but then that's lots of change , mainly in terms of tensorflow version. Trying to get this resolved with 0.4.1 release itself.

Sorry, but we explicitely stated on 0.4.1 release this was **experimental**.

**Comment by lissyx** [comment]
@rpratesh It looks like I do reproduce it on current master with TFLite on Linux and not just Android.

**Comment by lissyx** [comment]
@rpratesh So there was indeed an issue, looks like the API will have to evolve to properly fix that. PR https://github.com/mozilla/DeepSpeech/pull/2126 includes a fix but that will not be cool for thread-safety. You should be able to trivially apply that fix to 0.4.1 and rebuild with it.
